singularity-eos/base/fast-math/logs.hpp

```diff
@@ -78,76 +62,166 @@
 namespace singularity {
 namespace FastMath {
 
-// Integer Aliased versions. Likely not as portable
-// TODO(JMM): Add single-precision versions
+template <typename T>
+PORTABLE_FORCEINLINE_FUNCTION auto sgn(const T &x) {
+  return (T(0) < x) - (x < T(0));
+}
+
 PORTABLE_FORCEINLINE_FUNCTION
-auto as_long_int(double f) { return *reinterpret_cast<long long int *>(&f); }
+auto as_int(double f) { return *reinterpret_cast<std::int64_t *>(&f); }
 PORTABLE_FORCEINLINE_FUNCTION
-auto as_double(long long int i) { return *reinterpret_cast<double *>(&i); }
+auto as_double(std::int64_t i) { return *reinterpret_cast<double *>(&i); }
 
+// First order interpolation based NQTs
+// ----------------------------------------------------------------------
+// Reference implementations, however the integer cast implementation
+// below is probably faster.
 PORTABLE_FORCEINLINE_FUNCTION
-double fastlg_aliased(const double x) {
+double lg_o1_portable(const double x) {
+  int e;
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
+  const double m = frexp(x, &e);
+  return 2 * (m - 1) + e;
+}
+
+PORTABLE_FORCEINLINE_FUNCTION
+double pow2_o1_portable(const double x) {
+  const int flr = std::floor(x);
+  const double remainder = x - flr;
+  const double mantissa = 0.5 * (remainder + 1);
+  const double exponent = flr + 1;
+  return ldexp(mantissa, exponent);
+}
+
+// Integer aliased versions
+PORTABLE_FORCEINLINE_FUNCTION
+double lg_o1_aliased(const double x) {
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
   // Magic numbers constexpr because C++ doesn't constexpr reinterpret casts
   // these are floating point numbers as reinterpreted as integers.
-  // as_long_int(1.0)
-  constexpr long long int one_as_long_int = 4607182418800017408;
-  // 1./static_cast<double>(as_long_int(2.0) - as_long_int(1.0))
+  // as_int(1.0)
+  constexpr std::int64_t one_as_int = 4607182418800017408;
```

Comment:
This is used a few times, can it be placed in a higher scope so that it's not redefined in several places? Same for a couple other constants used.

Reply:
Done